### PR TITLE
Export optimization idea

### PR DESCRIPTION
**Pull request by benrudolph**
@dimagi/scale-team I'm trying to find a way to optimize the new export code so that giant apps like ICDS can build exports. It's challenging because the applications for ICDS are huge. I did some initial profiling and I get the sense that a lot of the issues come from searching through the giant couch doc to get properties as well as wrapping the huge doc. See profiling data on this ticket if curious (I also have more locally) http://manage.dimagi.com/default.asp?227649

Originally I was thinking of optimizing the how we get case properties, but that seemed like a rat's nest. So I decided to try and fetch less apps. I came up with two strategies to do this, and curious to hear if you have any feedback:

**Idea 1** (https://github.com/dimagi/commcare-hq/commit/2a7d4b4af8fe9ec88f953bfabd4eb514fe4df21c):
Mark all builds with a flag that says whether or not it has had form submissions. Then when we fetch apps during the export process we only fetch those that have had data submitted against them

Pros:
- More aggressive optimization and less apps pulled down (potentially only a few)

Cons:
- Adds logic to our form processing code

**Idea 2** (https://github.com/dimagi/commcare-hq/commit/942de756b84e1748506c05ab17f29f44baf81805):
Mark builds that have been marked as released at least one time (so that there is the possibility of data submissions)

Pros:
- Doesn't touch our form processing code

Cons:
- Could falsely analyze apps that have had no data submitted against them
- Seems like this could actually miss potential apps that have had data submitted to them

### CODE REVIEW

**Comment by benrudolph** [comment]
cc: @millerdev

File: corehq/apps/app_manager/models.py [hunk 1]
```diff
@@ -4781,6 +4781,9 @@ def validate_detail_screen_field(field):
 
 class SavedAppBuild(ApplicationBase):
 
+    # Whether the application has had data submitted against it
+    has_submissions = BooleanProperty(default=None)
```
Comment (benrudolph) [inline corehq/apps/app_manager/models.py]: default should actually be false and if the property is missing then we consider it "unknown" and analyze it. this avoids a migration

**Comment by snopoke** [comment]
Not a huge fan of having to load the app for every form submission. Is the issue that generating the schema requires iterating through all the builds and we're doing that when we load the UI?

Would another option be to update the schema when and app is released using the `app_post_release` signal?

**Comment by benrudolph** [comment]
I am referring to iterating over builds during the UI. Yeah I agree though it's easily cacheable for a long period of time, since we're talking about SavedApps which are immutable, and it can happen asynchronously.

We rebuild the schema from scratch every time we create a new export which arguably we don't have to do. See this flag (https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/export/models/new.py#L897). I'm hesitant to persist the schema throughout because if it gets in a weird state due to some bug, we're stuck telling everyone to have to hit some refresh schema button. It's been useful to have a new schema on each export, but that advantage may not be worthwhile anymore

**Comment by snopoke** [comment]
Ha, I was under the impression that we were persisting them. The code you
linked to does check for an existing schema to update (
https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/export/models/new.py#L906)
and it does also save the schema at the end (
https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/export/models/new.py#L941
).

Am I missing something about how that works?

**Comment by benrudolph** [comment]
Yeah so when we create a new export we intentionally set the `force_rebuild` flag (https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/export/views.py#L1471) which disregards the previous saved schema: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/export/models/new.py#L907

**Comment by benrudolph** [comment]
(going pillow route as discussed in meeting. thanks for the feedback)

**Pull request closed without merging**